File: Strings/3. Valid Anagram.js
//Leetcode: 242: Valid Anagram: https://leetcode.com/problems/valid-anagram/description/

// Given two strings s and t, return true if t is an 
// anagram of s, and false otherwise.

// Example 1:
// Input: s = "anagram", t = "nagaram"
// Output: true

 //Approach: O(nlogn)
 //Step1: split the string into an array and sort them.
 //Step2: compare each element and check if they are equal.
 //Step3: return  true or false based on that.

 var isAnagram = function(s, t) {
    if(s.length != t.length) return false;

    if(s.split('').sort().join() === t.split('').sort().join()){
        return true
    }else{
        return false
    }
};

File: 13. String/3. ValidAnagram.py
# Given two strings s and t, return true if t is an anagram of s, and false otherwise.

# Example:
# Input: s = "anagram", t = "nagaram"
# Output: true

# Brute Force: T: O(nlogn) | S: O(1)
# Step1: Convert string to array of character.
# Step2: Sort the arrays.
# Step3: Check if array elements are same.

def isAnagram(s: str, t: str) -> bool:
    if (len(s) != len(t)):
        return False
    
    sorted_s = list(s)
    sorted_t = list(t)
    sorted_s.sort()
    sorted_t.sort()

    for i in range(len(sorted_s)):
        if(sorted_s[i] != sorted_t[i]):
            return False
    
    return True

# Optimal Approach: T: O(n) | S: O(n)
# Step1: Create frequency map of both strings.
# Step2: Compare both the frequency maps. (The comparison only works for dict and set since they are unordered)
def isAnagram(s: str, t: str) -> bool:
    if(len(s) != len(t)):
        return False
        
    trackS = {}
    trackT = {}

    for i in range(len(s)):
        if s[i] in trackS:
            trackS[s[i]] += 1
        else:
            trackS[s[i]] = 1
    
    for i in range(len(t)):
        if t[i] in trackT:
            trackT[t[i]] += 1
        else:
            trackT[t[i]] = 1

    return trackS == trackT